File: src/domain/dyno-csv-parser.js
// dyno-csv-parser — lê CSV de dinamômetro (Dynojet/Mustang/Dynapack/genérico)
// e devolve uma curva normalizada `{ format, points: [{rpm, torque_nm, power_kw}] }`.
//
// Estratégia: detecta o separador, identifica colunas por regex no header,
// converte unidades (lb-ft → Nm, hp/cv → kW) e completa torque↔potência via
// a relação P(kW) = T(Nm) × RPM / 9549.

const HP_TO_KW = 0.7457;
const LBFT_TO_NM = 1.355818;

function detectSeparator(line) {
  const candidates = [',', ';', '\t'];
  let best = ',', bestCount = 0;
  for (const c of candidates) {
    const re = new RegExp(c === '\t' ? '\\t' : c.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&'), 'g');
    const cnt = (line.match(re) || []).length;
    if (cnt > bestCount) { bestCount = cnt; best = c; }
  }
  return best;
}

function detectFormat(headersStr) {
  const h = headersStr.toLowerCase();
  if (h.includes('dynojet')) return 'dynojet';
  if (h.includes('mustang')) return 'mustang';
  if (h.includes('dynapack')) return 'dynapack';
  return 'generic';
}

function findCol(headers, patterns) {
  for (let i = 0; i < headers.length; i++) {
    const h = headers[i].toLowerCase();
    for (const p of patterns) {
      if (p.test(h)) return i;
    }
  }
  return -1;
}

function parseNum(s) {
  if (s == null) return NaN;
  const cleaned = String(s).replace(/\s/g, '').replace(/,(?=\d{3}\b)/g, '').replace(',', '.');
  const v = parseFloat(cleaned);
  return Number.isFinite(v) ? v : NaN;
}

export function parseDynoCsv(text) {
  if (typeof text !== 'string' || text.trim().length === 0) {
    throw new Error('CSV vazio ou inválido');
  }
  const allLines = text.split(/\r?\n/).map(l => l.trim()).filter(l => l.length > 0);
  if (allLines.length < 2) {
    throw new Error('CSV precisa de header + ao menos 1 linha de dados');
  }
  // CSVs reais frequentemente têm linhas de metadata antes do header.
  // Header = primeira linha que contém a palavra "rpm" (case-insensitive).
  let headerIdx = -1;
  for (let i = 0; i < allLines.length; i++) {
    if (/\brpm\b/i.test(allLines[i]) && /[,;\t]/.test(allLines[i])) {
      headerIdx = i;
      break;
    }
  }
  if (headerIdx < 0) headerIdx = 0;
  const preamble = allLines.slice(0, headerIdx).join(' ');
  const lines = allLines.slice(headerIdx);
  const sep = detectSeparator(lines[0]);
  const headers = lines[0].split(sep).map(h => h.trim());
  const format = detectFormat(preamble + ' ' + lines[0]);

  const idxRpm = findCol(headers, [/^rpm$/i, /\brpm\b/i, /engine\s*rpm/i, /eng\.?\s*rpm/i]);
  if (idxRpm < 0) throw new Error('Coluna de RPM não encontrada');

  const idxTorqueNm  = findCol(headers, [/torque.*\bnm\b/i, /torq.*\bnm\b/i, /torque.*\(n[·\.\s\-]*m\)/i]);
  const idxTorqueLb  = idxTorqueNm < 0 ? findCol(headers, [/torque.*lb[\s\-]?ft/i, /torq.*lb[\s\-]?ft/i, /torque.*ftlb/i]) : -1;
  const idxTorqueGen = (idxTorqueNm < 0 && idxTorqueLb < 0) ? findCol(headers, [/^torque$/i, /\btorque\b/i, /\btorq\b/i]) : -1;

  const idxPowerKw   = findCol(headers, [/power.*\bkw\b/i, /pot.*\bkw\b/i, /\(kw\)/i]);
  const idxPowerHp   = idxPowerKw < 0 ? findCol(headers, [/power.*\bhp\b/i, /\bhp\b/i, /\bbhp\b/i]) : -1;
  const idxPowerCv   = (idxPowerKw < 0 && idxPowerHp < 0) ? findCol(headers, [/\bcv\b/i]) : -1;
  const idxPowerGen  = (idxPowerKw < 0 && idxPowerHp < 0 && idxPowerCv < 0) ? findCol(headers, [/^power$/i, /\bpower\b/i, /\bpot[eê]ncia\b/i]) : -1;

  if (idxTorqueNm < 0 && idxTorqueLb < 0 && idxTorqueGen < 0
      && idxPowerKw < 0 && idxPowerHp < 0 && idxPowerCv < 0 && idxPowerGen < 0) {
    throw new Error('Nenhuma coluna de torque ou potência encontrada');
  }

  const points = [];
  for (let li = 1; li < lines.length; li++) {
    const cells = lines[li].split(sep).map(c => c.trim());
    if (cells.length < 2) continue;
    const rpm = parseNum(cells[idxRpm]);
    if (!Number.isFinite(rpm) || rpm <= 0) continue;

    let torque_nm = null;
    if (idxTorqueNm >= 0) torque_nm = parseNum(cells[idxTorqueNm]);
    else if (idxTorqueLb >= 0) {
      const v = parseNum(cells[idxTorqueLb]);
      if (Number.isFinite(v)) torque_nm = v * LBFT_TO_NM;
    } else if (idxTorqueGen >= 0) {
      torque_nm = parseNum(cells[idxTorqueGen]);
    }

    let power_kw = null;
    if (idxPowerKw >= 0) power_kw = parseNum(cells[idxPowerKw]);
    else if (idxPowerHp >= 0) {
      const v = parseNum(cells[idxPowerHp]);
      if (Number.isFinite(v)) power_kw = v * HP_TO_KW;
    } else if (idxPowerCv >= 0) {
      const v = parseNum(cells[idxPowerCv]);
      if (Number.isFinite(v)) power_kw = v * HP_TO_KW; // cv ≈ 0.7457 kW
    } else if (idxPowerGen >= 0) {
      power_kw = parseNum(cells[idxPowerGen]);
    }

    if (Number.isFinite(torque_nm) && !Number.isFinite(power_kw)) {
      power_kw = torque_nm * rpm / 9549;
    }
    if (!Number.isFinite(torque_nm) && Number.isFinite(power_kw)) {
      torque_nm = power_kw * 9549 / rpm;
    }

    if (!Number.isFinite(torque_nm) && !Number.isFinite(power_kw)) continue;

    points.push({
      rpm,
      torque_nm: Number.isFinite(torque_nm) ? torque_nm : null,
      power_kw: Number.isFinite(power_kw) ? power_kw : null
    });
  }

  if (points.length < 3) {
    throw new Error('Curva precisa de ao menos 3 pontos válidos (encontrei ' + points.length + ')');
  }
  points.sort((a, b) => a.rpm - b.rpm);
  return { format, points };
}

File: ios/p1fast-core/Sources/P1FastCore/DynoCsvParser.swift
// ═══════════════════════════════════════════════════════════
// DynoCsvParser — port Swift de src/domain/dyno-csv-parser.js
// ═══════════════════════════════════════════════════════════
// Lê CSV de dinamômetro (Dynojet/Mustang/Dynapack/genérico) e devolve
// uma curva normalizada (`format`, `points: [DynoPoint]`).
//
// Detecta separador (vírgula/ponto-e-vírgula/tab), identifica colunas
// por regex no header, converte unidades (lb-ft → Nm, hp/cv → kW) e
// completa torque ↔ potência via P(kW) = T(Nm) × RPM / 9549.
//
// MS-9.1 (port 2026-05-04). Usado pelo cadastro de carro (cadastro
// de dyno_curve em MS-9 quando T4000 entra).

import Foundation

public enum DynoFormat: String, Sendable, Codable {
    case dynojet
    case mustang
    case dynapack
    case generic
}

public struct ParsedDynoCurve: Sendable, Equatable {
    public let format: DynoFormat
    public let points: [DynoPoint]
}

public enum DynoCsvParserError: Error, Equatable, Sendable {
    case empty
    case headerNotFound
    case rpmColumnNotFound
    case noTorqueOrPowerColumn
    case insufficientPoints(found: Int)
}

public enum DynoCsvParser {
    public static let hpToKw: Double      = 0.7457
    public static let lbftToNm: Double    = 1.355818
    public static let powerKwFromTorque: Double = 9549  // P(kW) = T(Nm) × RPM / 9549

    /// Parse principal. Lança em CSV inválido / sem RPM / < 3 pontos.
    public static func parse(_ text: String) throws -> ParsedDynoCurve {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw DynoCsvParserError.empty
        }
        let allLines = text
            .components(separatedBy: CharacterSet(charactersIn: "\n\r"))
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }
        guard allLines.count >= 2 else { throw DynoCsvParserError.empty }

        // Header = primeira linha que contém "rpm" + um separador.
        var headerIdx: Int = -1
        let rpmRegex = try NSRegularExpression(pattern: #"\brpm\b"#, options: [.caseInsensitive])
        let sepCharSet = CharacterSet(charactersIn: ",;\t")
        for (i, l) in allLines.enumerated() {
            let nsl = l as NSString
            let hasRpm = rpmRegex.firstMatch(in: l, range: NSRange(location: 0, length: nsl.length)) != nil
            let hasSep = l.rangeOfCharacter(from: sepCharSet) != nil
            if hasRpm && hasSep { headerIdx = i; break }
        }
        if headerIdx < 0 { headerIdx = 0 }
        let preamble = allLines[..<headerIdx].joined(separator: " ")
        let lines = Array(allLines[headerIdx...])

        let sep = detectSeparator(lines[0])
        let headers = lines[0]
            .components(separatedBy: String(sep))
            .map { $0.trimmingCharacters(in: .whitespaces) }
        let format = detectFormat(preamble + " " + lines[0])

        let idxRpm = findCol(headers, patterns: [
            #"^rpm$"#, #"\brpm\b"#, #"engine\s*rpm"#, #"eng\.?\s*rpm"#
        ])
        guard idxRpm >= 0 else { throw DynoCsvParserError.rpmColumnNotFound }

        let idxTorqueNm = findCol(headers, patterns: [
            #"torque.*\bnm\b"#, #"torq.*\bnm\b"#, #"torque.*\(n[·\.\s\-]*m\)"#
        ])
        let idxTorqueLb = idxTorqueNm < 0
            ? findCol(headers, patterns: [#"torque.*lb[\s\-]?ft"#, #"torq.*lb[\s\-]?ft"#, #"torque.*ftlb"#])
            : -1
        let idxTorqueGen = (idxTorqueNm < 0 && idxTorqueLb < 0)
            ? findCol(headers, patterns: [#"^torque$"#, #"\btorque\b"#, #"\btorq\b"#])
            : -1

        let idxPowerKw = findCol(headers, patterns: [#"power.*\bkw\b"#, #"pot.*\bkw\b"#, #"\(kw\)"#])
        let idxPowerHp = idxPowerKw < 0
            ? findCol(headers, patterns: [#"power.*\bhp\b"#, #"\bhp\b"#, #"\bbhp\b"#])
            : -1
        let idxPowerCv = (idxPowerKw < 0 && idxPowerHp < 0)
            ? findCol(headers, patterns: [#"\bcv\b"#])
            : -1
        let idxPowerGen = (idxPowerKw < 0 && idxPowerHp < 0 && idxPowerCv < 0)
            ? findCol(headers, patterns: [#"^power$"#, #"\bpower\b"#, #"\bpot[eê]ncia\b"#])
            : -1

        if idxTorqueNm < 0 && idxTorqueLb < 0 && idxTorqueGen < 0
            && idxPowerKw < 0 && idxPowerHp < 0 && idxPowerCv < 0 && idxPowerGen < 0 {
            throw DynoCsvParserError.noTorqueOrPowerColumn
        }

        var points: [DynoPoint] = []
        for li in 1..<lines.count {
            let cells = lines[li]
                .components(separatedBy: String(sep))
                .map { $0.trimmingCharacters(in: .whitespaces) }
            if cells.count < 2 { continue }
            guard idxRpm < cells.count else { continue }
            guard let rpm = parseNum(cells[idxRpm]), rpm.isFinite, rpm > 0 else { continue }

            var torqueNm: Double? = nil
            if idxTorqueNm >= 0, idxTorqueNm < cells.count {
                torqueNm = parseNum(cells[idxTorqueNm])
            } else if idxTorqueLb >= 0, idxTorqueLb < cells.count {
                if let v = parseNum(cells[idxTorqueLb]) { torqueNm = v * lbftToNm }
            } else if idxTorqueGen >= 0, idxTorqueGen < cells.count {
                torqueNm = parseNum(cells[idxTorqueGen])
            }

            var powerKw: Double? = nil
            if idxPowerKw >= 0, idxPowerKw < cells.count {
                powerKw = parseNum(cells[idxPowerKw])
            } else if idxPowerHp >= 0, idxPowerHp < cells.count {
                if let v = parseNum(cells[idxPowerHp]) { powerKw = v * hpToKw }
            } else if idxPowerCv >= 0, idxPowerCv < cells.count {
                if let v = parseNum(cells[idxPowerCv]) { powerKw = v * hpToKw }
            } else if idxPowerGen >= 0, idxPowerGen < cells.count {
                powerKw = parseNum(cells[idxPowerGen])
            }

            // Completar torque ↔ potência
            if let t = torqueNm, t.isFinite, !(powerKw?.isFinite ?? false) {
                powerKw = t * rpm / powerKwFromTorque
            }
            if let p = powerKw, p.isFinite, !(torqueNm?.isFinite ?? false) {
                torqueNm = p * powerKwFromTorque / rpm
            }

            // Pelo menos um dos dois precisa estar definido
            let tValid = torqueNm?.isFinite ?? false
            let pValid = powerKw?.isFinite ?? false
            if !tValid && !pValid { continue }

            points.append(DynoPoint(
                rpm: rpm,
                powerKw: powerKw ?? 0
            ))
            // Note: nosso DynoPoint só carrega rpm + powerKw. Torque-nm
            // é apenas intermediário na conversão. Se necessário pra
            // análise, usar p * 9549 / rpm.
        }

        if points.count < 3 {
            throw DynoCsvParserError.insufficientPoints(found: points.count)
        }
        points.sort { $0.rpm < $1.rpm }
        return ParsedDynoCurve(format: format, points: points)
    }
}

// MARK: - Internals

private func detectSeparator(_ line: String) -> Character {
    let candidates: [Character] = [",", ";", "\t"]
    var best: Character = ","
    var bestCount = 0
    for c in candidates {
        let cnt = line.reduce(0) { $0 + ($1 == c ? 1 : 0) }
        if cnt > bestCount { bestCount = cnt; best = c }
    }
    return best
}

private func detectFormat(_ s: String) -> DynoFormat {
    let lower = s.lowercased()
    if lower.contains("dynojet")  { return .dynojet }
    if lower.contains("mustang")  { return .mustang }
    if lower.contains("dynapack") { return .dynapack }
    return .generic
}

private func findCol(_ headers: [String], patterns: [String]) -> Int {
    for (i, h) in headers.enumerated() {
        let lower = h.lowercased()
        for pStr in patterns {
            if let re = try? NSRegularExpression(pattern: pStr, options: [.caseInsensitive]) {
                let nsl = lower as NSString
                if re.firstMatch(in: lower, range: NSRange(location: 0, length: nsl.length)) != nil {
                    return i
                }
            }
        }
    }
    return -1
}

private func parseNum(_ s: String) -> Double? {
    var cleaned = s.replacingOccurrences(of: " ", with: "")
    // remove vírgula como separador de milhar
    while let range = cleaned.range(of: #",(?=\d{3}\b)"#, options: .regularExpression) {
        cleaned.removeSubrange(range)
    }
    cleaned = cleaned.replacingOccurrences(of: ",", with: ".")
    return Double(cleaned).flatMap { $0.isFinite ? $0 : nil }
}
